tests/page-objects/forms/cht/pregnancy-form-factory.po.js

```diff
@@ -0,0 +1,33 @@
+const Factory = require('rosie').Factory;
+
+Factory.define('pregnancy-form')
+  .option('howWouldYouLikeToReportValue', 'method_approx')
+  .attr('howWouldYouLikeToReport',['howWouldYouLikeToReportValue'], (howWouldYouLikeToReportValue) => {
+    return {
+      css: 'input[name="/pregnancy/gestational_age/register_method/lmp_method"]',
+      value: howWouldYouLikeToReportValue,
+      endOfPage: true
```

Comment:
Commenting now to give some basic thoughts, but will add more thoughts later....

I really like the inclusion of rosie to make shared definitions of data structures that we need in a lot of places.

I wasn't expecting these factories to include anything other than data. The `css` and `endOfPage` feel like mixing concerns and limiting reusability beyond the e2e use case. That is signaled by the file name including `.po` in it, but what if these rosie factories were made to be useful for any test in the code (unit tests included)?

Also, it may not really hurt to have some extra fields in here now, but it does start to form an undesirable pattern, where we wouldn't want others to follow that mixing of concerns.

An alternative that comes to mind would just be to have a rosie factory and then some e2e function(s) that augment the factory-produced object with e2e concerns. That might easier said than done though and will need some thought.

Reply:
I think the `.po` was a mistake on my naming.  I don't see how this would be used in say unit tests as it is for UI form filling.  Come to think of it. We do something already in the medic-conf-test-harness but it isn't built to be used in an e2e test but does pretty much the same thing. 

This function takes a list of answers and then the harness finds the questions to fill. 
https://github.com/medic/cht-core/blob/aa9efbd6b7b7b98df2b98d4a2461a81ef5d2fcfd/config/default/test/forms/pregnancy.spec.js#L27

A list of answers. Pregnancies with different results. 

https://github.com/medic/cht-core/blob/master/config/default/test/form-inputs.js

Reply:
I think I found a way for this to feel “clean” and reusable (not tied to forms) but still be useful for filling in forms, since that is a big use case.

First, some background and what I think is valuable from a reuse perspective when it comes to object factories: We have cases around the codebase that need data to exercise behavior. One such example can be seen in `webapp/tests/karma/ts/services/extract-lineage.service.spec.ts:30`. In that test we set up the following object:
```
    const contact = {
      _id: 'a',
      name: 'jim',
      parent: {
        _id: 'b',
        age: 55,
        parent: {
          _id: 'c',
          sex: true,
          parent: {
            _id: 'd'
          }
        }
      }
    };
```

That is just a manually coded object, where someone is responsible for getting it just right and of course everyone afterwards is expected to be careful when touching it. It doesn't come with any logical "schema" or general agreement that it's shaped according to anything. Grepping around the codebase for `const contact = ` can give some other examples. We tend to very sparsely populate those objects since most cases don't need anything fancy... or do they??!!

The idea of an object factory is to make it super easy to generate correct objects, where that correctness is handled in a single place (the rosie factory in this case) and trusted to always generate nice objects. That reduces the risk or someone throwing together an object, failing to realize they got it wrong, and then coding to their test, which is now also wrong, possibly just for a special case that will slip by.

The other big thing here is that assumption that sparsely populated objects are enough. Often they are. Sometimes more data is needed to test a case. Sometimes the presence of more data will drive the need for handling a case, when otherwise it wouldn't have come up had the data been absent. I wouldn't ask everyone to make objects with all the fields filled in all the time. I would ask the computer to do it for us though!

I think most of this is already in line with your thoughts here, but the difference to this point is that I'm talking about unit and integration tests (not e2e). That's where the discussion about reusability comes in. My thought on object factories is that they generate data as it would be in the DB. That's not surprising from a read/query standpoint, but when we talk about forms and reports we need to address the input/save case. In that case, I still think it's useful to think of the data from a stored-doc perspective (which makes it useful everywhere else in the codebase for unit and integration tests) but augment it to address the very real need to map object fields to form fields.

Than "augment" part might just mean calling the regular factory to produce the data and then having a separate function that maps all the data fields to form fields. If Rosie provides additional value, and is used for the factory anyway, you can use the `extend` function/feature (see last line of [first code example](https://github.com/rosiejs/rosie#usage)) to then take that shared factory and add all the extra fields you want for the web/form needs.

So you'd end up with a `PregnancyReport` factory and then perhaps a `PregnancyReportForm` that extends `PregnancyReport` to add in whatever selector data and such that you might need for e2e tests.

Reply:
Yeah that makes sense. I'll give this another crack. 

Basically we'd end up with a object in test code from the factory that mimics what the report would get saved as from filling out the form. Then have some way to map the fields of the factory object to the UI fields. 

Makes sense.